README
[ 
  This is skank code from the dawn of time.  It's PHP.  It's mysql only.  It has
  0 tests.  Cut out your eyes.

  Just getting it off sourceforge and into github for posterity's sake.  The
  irony is that over 10 years after the original implementation and almost 9
  years after the last update, there are still running instances of this
  software.  Crazy. 
]

--------------------------------------------------------------------------------


Getting Started with RDTJ (Roundeye's Duct-Tape Jukebox)

by Rick Bradley (roundeye@roundeye.net / rick@eastcore.net)
Sep. 26, 2001

$Header: /cvsroot/rdtj/rdtj/README,v 1.6 2001/09/27 05:53:18 roundeye Exp $



  * * * * * * * * * * UPGRADE NOTICE * * * * * * * * * * * * * * * * * * * 

  NOTE -- if you're upgrading from 0.2 or 0.3 to 0.4 you need to non-destructively
  modify a database table.  After unpacking, feed the upgrade SQL file into
  your database, for example:

  % mysql -uuser -p -hhost mp3 < ~/rdtj-0.4/upgrades/tables_0.2,0.3-0.4.sql

  NOTE -- if you're upgrading from 0.1 you need to first non-destructively
  modify a database table.  After unpacking, feed the upgrade SQL file into
  your database, for example:

  % mysql -uuser -p -hhost mp3 < ~/rdtj-0.3/upgrades/tables_0.1-0.2.sql

  Then follow the above instructions on upgrading from 0.2 or 0.3 to 0.4.

  * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
  



About RDTJ

	"Roundeye's Duct-Tape Jukebox" is a web-based MP3 jukebox system.
	I wrote it prior to a party I was holding when I wanted to allow
	people to choose songs to play on the stereo without having to
	worry about whether they were too drunk to operate it...  I have a
	decent portion of my CD collection ripped to MP3 files which are
	organized by artist and album and browsing by artist and/or album
	seemed a really natural way to go about things.  The idea that
	partygoers (and later anyone over at the house and even others out
	on the web) could use a web browser to pick songs to play really
	appealed to me: no worries about whether the stereo would end up
	with beer in it, people don't have to wait around holding a CD
	while other songs finish, and since I have a bunch of computers
	around the house people don't need to go to another room to play a
	song.

	Since then some friends have installed the software and we have
	fun playing songs on each others stereos (even though some of us
	are states apart).  While I don't necessarily recommend making
	your stereo accessible from the WWW, using a firewall or an
	htaccess access control system to make it selectively available
	can make for some fun.


	The homepage for RDTJ is at http://rdtj.sourceforge.net



I - Getting the source

	There are currently two ways to get RDTJ:  HTTP download and CVS checkout.


					    HTTP download


	RDTJ is being hosted by SourceForge (a part of the Open Source
    Development Network - OSDN) which provides free hosting and
    development services for Open Source software (like RDTJ).  They
    offer HTTP (web) access to file repositories.

	To retrieve RDTJ version 0.4 via HTTP:

	Point a web browser at this address and save the resulting file:

		http://prdownloads.sourceforge.net/rdtj/rdtj-0.4.tar.gz
		   

      		
						CVS checkout


	If you have CVS installed you can check out the CVS source from
	Sourceforge via their CVS pserver.

	To check out the RDTJ CVS tree via pserver:

		% cvs -d:pserver:anonymous@cvs.rdtj.sourceforge.net:/cvsroot/rdtj login 

	[ when prompted for a password for "anonymous" simply press the Enter key ]
                   
        % cvs -z3 -d:pserver:anonymous@cvs.rdtj.sourceforge.net:/cvsroot/rdtj co rdtj




II - Prerequisites

	The RDTJ software is being developed on Linux, but will most
	likely work on a wide range of *nix systems (FreeBSD, OpenBSD,
	NetBSD, SunOS/Solaris, OSX, etc.) as long as they can run the
	necessary underlying packages.  While the front-end should run
	anywhere PHP (w/ MySQL support) runs, the back-end relies on
	Perl's fork() call to allow for real-time song skipping and hence
	won't run on Windows.

	You will need a number of packages to make the RDTJ software work
	properly:

	- a database (MySQL is the only one supported at the moment).  

		Currently you'll need MySQL installed somewhere on your
		network and the permissions set to allow the web server to
		perform selects, inserts, updates, and deletes on a database
		named "mp3".

	- a PHP-enabled WWW server.  

		I developed the code using Apache 1.3.*.  I had mod_php3
		compiled in and had MySQL support compiled into PHP3.  The
		system works fine (and is even faster) under PHP4.

        In future releases we will not guarantee PHP3 support.

	- Perl with the following non-standard module dependencies:

       ?  RDTJ component      Perl modules needed
       ----------------------------------------------------------------------
       M  jukebox_player.pl   DBI, DBD::mysql
       O  insert_album        MP3::Info
       O  add                 MP3::Info, LWP::Simple, Algorithm::Diff,
                              DBI, DBD::mysql       

       ? -> 'M' denotes mandatory components, 'O' denotes optional components

		I am using Perl 5.6.1 for development, but you can probably
		get away with most Perl releases > 5.004 (and in a pinch you
		might be able to go back further).

	- a command-line MP3 player.

		I use mpg123 (http://www.mpg123.de/).  It's slick, sturdy, and
		does the job.  You could probably find a number of other
		players which will fit the bill.  Currently if you change
		players you may need to do minor code hacking in
		jukebox_player.pl.  This hassle should be alleviated in a
		future release.

	- a system with sound support which can play mp3's.

		I took a laptop with a line out jack, hooked up some RCA
		cables into the back of my stereo (the AUX jack for those who
		must know) and when the laptop plays so does the stereo.


III - Installing.

	You can install the system on one machine (which is probably the
	easiest way to go for starters) or spread the components over
	different machines.  For the sake of the installation discussion
	assume you have the following machines:

		- dbserver   - a computer which is running MySQL
		- wwwserver  - a computer running a PHP-enabled web server
		- playserver - a computer with a sound card which will run the
		               backend player and mpg123

	Once you have the components listed above in place you need to
	unpack and install RDTJ.

	- Move the downloaded rdtj-0.4.tar.gz to a working directory.
	- Unpack rdtj by issuing the command:

		% tar xvfz rdtj-0.4.tar.gz
		% cd rdtj-0.4/

	- Create a directory in the wwwserver's web documents directory
	for the rdtj php interface and copy the files in rdtj-0.4/html/
	there:

	% mkdir /home/httpd/html/rdtj/
	% cp -a html/* /home/httpd/html/rdtj/

	or:

	% cp -a html/  /home/httpd/html/rdtj/

	- update the front-end configuration

	The front-end configuration is now much more flexible, more
	centralized, and hopefully easier to understand.  The file 
	rdtj-0.4/html/header.php contains the configuration options (as
	well as the header for the web pages).  Here is how the
	configuration section looks by default:


    # ADMIN OPTIONS
    $admin_pass		= "pass";		# password for administrative functions
									# (should be md5+database eventually)
    $admin_allow_song	= 1;		# allow admin to modify song details?
    $admin_allow_album  = 1;		# allow admin to modify album details?
    $admin_allow_artist = 1;		# allow admin to modify artist details?

    # DATABASE OPTIONS

    $dbhost = "host";				# which host is the database server?
    $dbuser = "user";				# which user should connect to the db?
    $dbpass = "password";			# what is the db user's password?
    $dbname = "mp3";				# what is the name of the database?

    # COLORS
    $color_page		= "#f2f2d0";	# background color for page
    $color_request	= "#f9fce8";	# background color for requests area
    $color_status	= "#71b8f2";	# background color for current song area
    $color_controls	= "#a2d2f2";	# background color for player controls
    $color_hits		= "#bffce6";	# background color for hits area
    $color_songs	= "#a1f2f4";	# background color for collapsed song list
    $color_text		= "#000000";	# default text color
    $color_error_page	= "#f4f44b";	# background color for error sections
    $color_error_msg	= "#f21515";	# background color for error messages

    # STATUS OPTIONS
    $status_title	= "Jukebox Status"; # headline to show above status area
    $status_album	= 1;			# display album for current song?
    $status_length	= 1;			# show song length for current song?
    $status_show_IP	= 1;			# show IP for current song?
    $status_play_line	= "Now playing:";    # "now playing" message text

    # PLAYER CONTROL OPTIONS
    $controls_on	= 1;			# enable player controls?
    $controls_show_IP	= 1;		# show IP in player controls?
    $controls_show_time = 1;		# show elapsed time in player controls?

    # REQUEST OPTIONS
    $request_title	= "Requests";	# headline to show above expanded requests
    $request_album	= 1;			# display albums in request list?
    $request_length	= 1;			# show track length in request list?
    $request_show_IP	= 1;		# show IP's in request list?
    $request_wait_time	= 1;		# show "requested xxx ago" info for requests?
    $request_kill	= 1;			# display "kill" links in requests?
    $request_close	= 1;			# "close" link on the expanded requests area?
    $mass_murder	= 1;			# display mass-delete widgets in requests?

    # HITS OPTIONS
    $hits_title		= "Greatest Hits";  # headline to show above expanded hits
    $hits_default	= 10;			# default number of hits to show initially
    $hits_all		= 1000000;		# number of hits to request for "all hits"
    $hits_close		= 1;			# "close" link on the expanded hits area?

    # SONG LIST OPTIONS
    $songs_title	= "";			# headline to show above song list

    # MISC. OPTIONS
    $refresh_on		= 1;			# should the page refresh?
    $refresh_time	= 150;			# how many seconds before a refresh?
    $refresh_time_small = 30;		# refresh time when song list is hidden



	    You'll at least want to change $admin_pass to be a secret
	    administrator password.  You'll also need to set $dbhost
	    ("localhost" will probably work if you are running everything on
	    one machine), $dbuser, and $dbpass.

	    To enable administration you will also need to set
	    $admin_allow_song, $admin_allow_album, and $admin_allow_artist
	    to 1.


	- Create the mp3 database and load the tables into it

	% echo "create database mp3" | mysql -uusername -p -hdbserver 
	% mysql -uusername -p -hdbserver < tables.sql


	- Configure the Perl components

	The Perl components of the RDTJ share a common configuration file
    located at rdtj-0.4/RDTJ/Config.pm  You should edit this file's
    settings to reflect your installation.  You will at least want
    to update the database settings:

	# database settings
	$dbname = "DBI:mysql:mp3";
	$dbuser = "user";
	$dbpass = "password";

	To connect on another host you'll want to use something like:

	$dbname = "DBI:mysql:mp3@dbserver:port";

	There is additionally a log file which is created by jukebox_player.pl.
	By default the output goes to the standard output (stdout) which is
	specified by '-'.  If you want to log to a file then set $logfile
	to point to where you want your logfile to live.  The player will
	close the log file, time stamp it, and open a new one if you send it
	a HUP signal.


	- Load up your mp3 data into the database

	This is currently the most "duct-tape-ish" part of the
	installation (you knew the name must have come from somewhere,
	right?).  Fortunately it's also the area I'm most excited

CHANGELOG
CHANGELOG for RDTJ (Roundeye's Duct-Tape Jukebox)

See LICENSE file for license and copying information.

[ times are GMT ]

[ RDTJ 0.4 released ]

date: 2001/09/28 05:56:06 author:roundeye
- fixed another db busy-waiting loop bug

date: 2001/09/27 20:24:30 author:roundeye
- decreased mp3 player verbosity
- fixed bug with non-playable files causing busy-waiting db loop

date: 2001/09/27 06:39:31 author:roundeye
- minor tweaks to random song logic after self-audit of code

[ 0.4 Release Candidate 1]
date: 2001/09/27 05:53:18 author:roundeye
- updated README for 0.4
- fixed minor usage string error in jukebox_player.pl

date: 2001/09/27 01:52:09 author:roundeye
- rewrote main back-end play loop
  - fixed timing issues with random songs and MySQL
  - streamlined
  - set reasonable query delays
- updated front-end color settings
- turned off back-end debugging by default
- dropped use of player tag in human-readable front-end UI (use only player message)
- miscellaneous cleanup

date: 2001/09/26 05:47:47 author:roundeye
- merged vigilante code with jukebox code (we now use a fork() model)
- added a SIGCHLD handler
- factored database code out to RDTJ::DB
- factored database queries out into hash structure
  (significantly trimmed redundant db code)
- added hack to allow RDTJ::Config to be read from PHP as well :-)

date: 2001/09/23 19:54:41 author: roundeye
- factored out Perl logging code to RDTJ::Log
- factored out Perl configuration code to RDTJ::Config
- factored out Perl signal handlers to RDTJ::RDTJ
- now initialize Perl vars from RDTJ::Config values
- updating file headers (license info, CHANGELOGs, etc.)
- removed CVS Log headers
- cleaned up formatting/indentation in numerous places
- fixed vigilante version of PlayerName/SafePlayerName bug
- now use FindBin to locate vigilante (and new RDTJ:: libs)

date: 2001/09/21 04:27:50;  author: roundeye
- fixed: /index.php URLS -> index.php (patch submitted by Ryan Drake <ryan@mediaone.net>)
- fixed: tables.sql queue table now includes 'player' column (patch submitted by Ryan Drake <ryan@mediaone.net>)
- jukebox_player.pl: minor tweaks to random song routine (variable rename, added sleep 1) --
  this is only a temporary fix until tomorrow :-)

date: 2001/08/11 18:44:09;  author: roundeye
- jukebox_player.pl: Added support for random songs when playlist(s) empty
- index.php: toggle switch for same
- index.php: added total music time to request list display (useful for making cassette
  tapes, e.g.)

date: 2001/07/20 22:24:44;  author: roundeye
- jukebox_player.pl: er.  I mean fixed the previous bug (infinite quote escaping)

date: 2001/07/20 22:19:09;  author: roundeye
- jukebox_player.pl: fixed "infinite quote-escaping" bug when player gets disconnected

date: 2001/01/09 11:21:08;  author: roundeye
- jukebox_player.pl: now support migration of requests from one player to another :-)

date: 2001/01/08 13:46:03;  author: roundeye
- jukebox_player.pl: shitheaded bug with jukebox<->vigilante interaction

date: 2001/01/08 13:23:21;  author: roundeye
- jukebox_player.pl: front-end now supports multiple jukeboxes

date: 2001/01/08 11:55:50;  author: roundeye
- update_rdtj now starts jukebox properly

date: 2001/01/08 11:54:35;  author: roundeye
- jukebox_player.pl: Player message now updating

date: 2001/01/08 11:47:53;  author: roundeye
- jukebox_player.pl: backend now supports multiple queues

date: 2001/01/08 11:02:11;  author: roundeye
- index.php is showing up modified, but a diff shows nothing; presumably this
  is a timestamp/clock issue
- moved upgrades/tables_1.1-1.2.sql to upgrades/tables_0.1-0.2.sql, which is
  what it should have been called in the first place
- added upgrades/tables_0.2,0.3-0.4.sql to allow migration between 0.[23] and 0.4
  which will include multi-player support

date: 2001/01/08 10:54:05;  author: roundeye
- added "force artist" and "force album" commands to add script

date: 2000/11/02 14:41:07;  author: roundeye
- tiny interface usability change -- when submitting songs, keep the
  proper expansion (artist expanded / album expanded) but on reload jump
  to the album from the submitted song (reduces scrolling for multi-album
  artists)

date: 2000/11/01 13:00:23;  author: roundeye
- wired 'add' up to the database
- added code to only add mp3's which are not already in the database.

date: 2000/10/24 13:26:18;  author: roundeye
- major work on "add" script to add mp3's to database

date: 2000/10/23 08:51:13;  author: roundeye
- add: adding more heuristics for name lookups
- add: added ability to search eastcore music database


date: 2000/10/16 20:48:50;  author: roundeye
- adds: adding heuristics: put in most common english words and
  english single-letter frequency data
- add: working on cleanup/final heuristics

date: 2000/10/16 19:59:39;  author: roundeye
- add: incomplete first check-in of heuristic mp3 insertion tool

[ RDTJ 0.3 released ]

date: 2000/10/16 13:33:46;  author: roundeye
- insert_album: fixed minor bug with insert_album:  it was intended to 
  be run for one album at a time, but it turns out that it works *almost* perfectly for
  multiple albums at once (perl insert_albums list1 list2 list3 ...)
  except for the fact that the album sort order field doesn't reset on
  album boundaries.  Sort order is still correct but instead of each album
  having tracks numbered 1, 2, 3... they continue on upwards.

date: 2000/10/16 13:06:43;  author: roundeye
- add update_rdtj, a little utility script to update rdtj 
  from CVS and restart player

date: 2000/10/16 12:37:33;  author: roundeye
- made 'ps' search more robust (was failing in some circumstances)

date: 2000/10/16 12:34:08;  author: roundeye
- jukebox_player.pl: added clean error page for db down situations
- jukebox_player.pl: workarounds for a couple of very annoying bugs in old mysql's

jukebox_player.pl
#!/usr/bin/perl -w

#
#  Back-end player for rdtj (Roundeye's duct-tape jukebox)
#
#  Rick Bradley - roundeye@roundeye.net / rick@eastcore.net
#
#  please consult LICENSE file for license information
#  CHANGELOG lists history and additional contributor information
#

#
# $Header: /cvsroot/rdtj/rdtj/jukebox_player.pl,v 1.22 2001/09/28 05:56:06 roundeye Exp $
#

use strict;			        # enforce some discipline
use FindBin;			    # find this binary (standard module)
use Getopt::Long;           # command-line argument processing (standard module)
use DBI;                    # generic database connection
use DBD::mysql;             # mysql specific drivers
use lib "$FindBin::Bin";	# find local perl libs
use RDTJ::RDTJ;			    # common RDTJ routines
use RDTJ::DB;			    # RDTJ database routines
use RDTJ::Config;           # get configuration information
use RDTJ::Log;              # logging facility

# declare our variables
use vars qw(
            $dbname $dbuser $dbpass $db %db_handles %queries $DEBUG $pid
            @lines $ps %ps_tree $ps_pid $ps_ppid $filename $qid $random $songid
            $logfile $mp3_player $play_buffer $dbsleep $nosongsleep $random_wait
            $PlayerName $PlayerMessage $logfile $row $random_row $seen 
           );

# perform command-line processing (also starts logging...)
process_cmdline();

#
# set signal handlers
#

$SIG{'__DIE__'}     = \&DIE_LOG;        # "die" messages are logged
$SIG{'__WARN__'}    = \&WARN_LOG;       # "warn" messages are logged
$SIG{'HUP'}         = \&HUP;            # rotate log file on SIGHUP
$SIG{'INT'}         = \&EXIT;           # ctrl-c kills, close log gracefully
$SIG{'KILL'}        = \&EXIT;           # SIGKILL kills, close log gracefully
$SIG{'CHLD'}        = \&REAPER;         # SIGCHLD wait()s and EXIT()s

#
# define all database queries (and associate with "handle" names)
#
%queries = (
            'register'      => 'replace into player set name=?, message=?',
            'next song'     => 'select action, filename, qid from songs, queue'
                               .' where player=? and action in (0,1) and status = 0'
                               .' and if(action=0, songid=song, songid=1)'
                               .' order by action desc,qid asc limit 1;',
            'mark playing'  => 'update queue set status=1 where qid=? and player=?',
            'done playing'  => 'update queue set status=2 where action = 0 and status = 1 and player=?',
            'skip playing'  => 'update queue set status=2 where qid=? and player=?',
            'random mode'   => 'select defaultaction from player where name=?',
            'random song'   => 'select songid from songs order by RAND() limit 1',
            'insert song'   => 'insert into queue (action, status, song, origin, player)'
                               ." values (0,0,?,'[random]', ?)",
            'find 86'       => 'select qid from queue where player=? and action=86 limit 1',
            'clear 86'      => 'delete from queue where action=86 and player=?',
#TODO: have this called prior to current fatal signal handlers...
            'clean up'      => 'delete from player where name=?',
           );

if ($pid = fork())
{
    # parent --> "vigilante" (kills player instances when instructed)

    LOG("Vigilante process pid = $$, player process pid = $pid");
 
    VIGDB:
    while (1)
    {
        # connect to database
        $db = db_connect($dbname, $dbuser, $dbpass) or do { sleep $dbsleep; next VIGDB; };
        LOG("vigilante: connected to database $dbname as user $dbuser");

        # register queries with database module
        db_register(\%queries, \%db_handles) or die "Cannot register database queries!\n";

        # clean up any old '86's still in the database
        db_execute('clear 86', $PlayerName) or do { sleep $dbsleep; next VIGDB; };

        VIGSONG:
        while (1)
        {
            # look for an '86' (kill player) command
            db_execute('find 86', $PlayerName) or do { sleep $dbsleep; next VIGDB; };

            $seen = 0;
            while ($row = $db_handles{'find 86'}->fetchrow_hashref)
            {
                $seen = 1;                          # we've seen an action

                # hunt down the current player
                { 
                    # disable SIGCHILD for this command...
                    local $SIG{'CHLD'} = 'IGNORE';
                    @lines = `$ps -falxwww`;
                }
                %ps_tree = ();                       # clear process tree hash
                foreach (@lines)
                {   # find the player's pid
                    next unless /$mp3_player/o;     # must be a player
                    # mpg123 runs a process (P) which spawns a child (C)
                    # to kill it immediately and cleanly you must 
                    # kill -INT C then kill -INT P

                    # get process information
                    (undef, undef, $ps_pid, $ps_ppid, undef) = split();
                    # store parent/child pid info 
                    $ps_tree{$ps_ppid} = $ps_pid;
                }

                # kill the player
                if (defined $ps_tree{$pid} and defined $ps_tree{$ps_tree{$pid}})
                {
                    LOG("killing $ps_tree{$ps_tree{$pid}}");
                    kill(9,$ps_tree{$ps_tree{$pid}}); 
                    LOG("killing $ps_tree{$pid}");
                    kill(9,$ps_tree{$pid});
                }
                else
                {
                    LOG("couldn't find process to kill (player = [$mp3_player])");
                }

                # clean up '86' entry
                db_execute('clear 86', $PlayerName) or do { sleep $dbsleep; next VIGDB; };
            }

            # don't busy-wait if there is no action
            sleep $nosongsleep unless $seen;
        }
    }
}
else
{
    # child --> player (plays songs)

    $SIG{'CHLD'} = 'IGNORE';        # verily you may kill mine children

    # query database for mp3's to play, play 'em, update queue
    DB:
    while (1)
    {   
        # connect to database
        $db = db_connect($dbname, $dbuser, $dbpass) or do { sleep $dbsleep; next DB; };
        LOG("connected to database $dbname as user $dbuser");

        # register queries with database module
        db_register(\%queries, \%db_handles) or die "Cannot register database queries!\n";

        # register this player
        # NOTE: contention checking is tough so we don't do any ;-P
        LOG("Registering player $PlayerName with message $PlayerMessage");
        db_execute('register', $PlayerName, $PlayerMessage) or do { sleep $dbsleep; next DB; };
        LOG("player $PlayerName successfully registered");

        # get rid of any orphaned "playing" tracks in queue
        db_execute('done playing', $PlayerName) or do { sleep $dbsleep; next DB; };
        LOG("Cleaned any old 'playing' tracks from queue for player $PlayerName");

        # repeatedly get and play a song
        SONG:
        while (1)
        {
            $seen = 0;

            # find the next song or action in our queue
            db_execute('next song', $PlayerName) or do { sleep $dbsleep; next DB; };

            if ($row = $db_handles{'next song'}->fetchrow_hashref)
            {   # song or action found
                $random_wait = 0;       # clear "waiting for random song" flag
                if (!$row->{'action'})
                {   # action == "play"
					$seen = 1;              			# we've seen a playable song
                    $qid = $row->{'qid'};               # get the queue id
                    $filename = $row->{'filename'};     # get the file name
                    if (-f $filename)
                    {	# file is there so try to play the song
                        LOG("Playing $filename");

                        # update "current track" status in queue
                        db_execute('mark playing', $qid, $PlayerName) or do { sleep $dbsleep; next DB; };

                        # actually play the mp3 song
                        # system($mp3_player, '-v', "-b $play_buffer", $filename); # verbose, but hell on logfiles
                        system($mp3_player, "-b $play_buffer", $filename);

                        # move song to "played" status in queue
                        db_execute('done playing', $PlayerName) or do { sleep $dbsleep; next DB; };
                    }
                    else
                    {   # skip this unplayable track
                        LOG('error', "Cannot play song [$filename]...  Skipping.");
                        db_execute('skip playing', $qid, $PlayerName) or do { sleep $dbsleep; next DB; };
                    }
                }
                else
                {   # non-"play" action
                }
            }
            else
            {   # no songs or actions
                # query the setting for random play mode
                db_execute('random mode', $PlayerName) or do { sleep $dbsleep; next DB; };
                $random = 0;
                while ($row = $db_handles{'random mode'}->fetchrow_hashref)
                {
                    $random = ($row->{'defaultaction'} == 1 ? 1 : 0);
                }

                if ($random)
                {   # random song mode is active
                    # check flag -- are we waiting for a random song to show up?
                    if ($random_wait and $random_wait < $RDTJ::Config::random_limit)
                    {   # waiting, but haven't waited long enough to give up
                        $random_wait++;     # track how long we wait
                        sleep 1;            # just wait a little...
                        next SONG;          # and check song list again
                    }

                    $random_wait = 0;

                    # pick a random song
                    db_execute('random song') or do { sleep $dbsleep; next DB; };

                    if(($random_row = $db_handles{'random song'}->fetchrow_hashref) and 
                       ($songid = $random_row->{'songid'}))
                    {
                        #  update queue with random song
                        db_execute('insert song', $songid, $PlayerName) or do { sleep $dbsleep; next DB; };

                        # initialize "waiting for random song" flag -- this
                        # addresses some annoying MySQL timing issues
                        $random_wait = 1;

                        # go back and check song list
                        next SONG;
                    }
                }
            }
            # don't busy-wait if there are no songs
            sleep $nosongsleep unless $seen;
        }
    }
}


# command line processing
sub process_cmdline
{
    my $prog = $0;                  # what's my name baby?
    $prog =~ s:.*?([^/]+)$:$1:;     # just a little trim job

    # allow CVS to keep track of versions and last updates :-)
    my $version = '(CVS revision #) $Revision: 1.22 $';     $version    =~ s/\$//g;
    my $lastupdate = '$Date: 2001/09/28 05:56:06 $';      $lastupdate =~ s/\$//g;

    my $usage = <<EOU;
Usage:  $0 -n playername [options]

Version $version last updated $lastupdate.

RDTJ - Roundeye's duct-tape jukebox.  This is the back-end player
script.  Install on a machine with a sound card (that hopefully has
access to your directories of mp3's, *and* has access to your database).
This program will connect to the database, and repeatedly pick the next
song from the queue, play it, and update the queue.

    General options


    -n, --name			            Name of this player (queue name on front-end)
    -m, --message		            Player description message
    -h, -?, --help                  Display this message
    -v, --version                   Output version info and exit
    -d, --debug                     Turn on debugging (default = off)
    -l file, --logfile file         Set log file to 'file'
                                    (use '-' for stdout)
    --nolog                         Disable logging

    --dbsleep seconds               # seconds to sleep if database down
    --qsleep  seconds               # seconds to sleep on empty queue
    --ps                            location of 'ps' binary

    Database options

    --db     name                   Name of database to connect to
    --dbhost hostname               Hostname to use for database connection
    --dbport port                   Port number to use for database connection
    --dbuser username               Username to use for database connection
    --dbpass password               Password to use for database connection
                                     (warning:  this is visible to 'ps'!)
    Player options
    
    -p path, --player path          Path to command-line mp3 player
                                    (default = /usr/bin/mpg123)
    -b size, --buffer size          Player buffer size in kilobytes
                                    (default = 1024)
                                    
Homepage: http://rdtj.sourceforge.net
Author:   Rick Bradley (roundeye\@roundeye.net / rick\@eastcore.net)

EOU

    # variables to store commandline args
    my ($help, $ver, $debug, $log_file, $nolog, $db_sleep, $q_sleep,
        $db_name, $db_host, $db_port, $db_user, $db_pass, $player, 
        $buffer, $name, $message, $local_ps);

    # don't ignore case -- in case we want to differentiate between '-x'/'-X'
    Getopt::Long::Configure('no_ignore_case');

    # retrieve the command line options
    GetOptions(
	       'name|n=s'	    => \$name,
	       'message|m=s'	=> \$message,
           'help|h|?'       => \$help,
           'version|v'      => \$ver,
           'debug|d'        => \$debug,
           'logfile|l=s'    => \$log_file,
           'nolog'          => \$nolog,
           'dbsleep=i'      => \$db_sleep,
           'qsleep=i'       => \$q_sleep,
           'db=s'           => \$db_name,
           'dbhost=s'       => \$db_host,
           'dbport=s'       => \$db_port,
           'dbuser=s'       => \$db_user,
           'dbpass=s'       => \$db_pass,
           'player|p=s'     => \$player,
           'buffer|b=s'     => \$buffer,
           'ps=s'           => \$local_ps,
    ) &&!$help &&!$ver or die $usage;

    die $usage if $ver;
    die $usage unless $name;			    # player name is mandatory

    # set defaults from RDTJ::Config
    $dbname 		= $RDTJ::Config::dbname;
    $dbuser 		= $RDTJ::Config::dbuser;
    $dbpass 		= $RDTJ::Config::dbpass;
    $dbsleep 		= $RDTJ::Config::dbsleep;
    $nosongsleep 	= $RDTJ::Config::nosongsleep;
    $mp3_player		= $RDTJ::Config::mp3_player;
    $play_buffer	= $RDTJ::Config::play_buffer;
    $DEBUG		    = $RDTJ::Config::DEBUG;
    $logfile		= $RDTJ::Config::logfile;
    $ps             = $RDTJ::Config::ps;
    
    # set local defaults
    $PlayerName = $name;			    # set global player name
    $PlayerMessage = $message || '';		    # set global player message

    # prepare for the inevitable
    my $dieflag = 0;
    my $dieerror = '';

    # process command-line arguments

    $DEBUG = $debug      if (defined $debug);       # set global DEBUG flag
    $logfile = $log_file if (defined $log_file);    # set global logfile

    if (defined $db_sleep)
    {   # validate positive integer
        if ($db_sleep <= 0)
        {
            $dieflag = 1;
            $dieerror .= "--dbsleep must specify a number > 0\n";
        }
        else
        {
            $dbsleep = $db_sleep;                   # set global $dbsleep
        }
    }

    if (defined $q_sleep)
    {   # validate positive integer
        if ($q_sleep < 0)
        {
            $dieflag = 1;
            $dieerror .= "--qsleep must specify a number >= 0\n";
        }
        else
        {
            $nosongsleep = $q_sleep;                # set global $nosongsleep
        }
    }

    if (defined $db_port and ($db_port < 0 or $db_port > 65535))
    {   # port number is out of range
        $dieflag = 1;
        $dieerror .= "--dbport must be between 0 and 65535\n";
    }

    if (defined $db_name or defined $db_host) 
    {   # must update the $dbname spec.  
        if (!defined $db_host) 
        {   # just the database name changed
            $dbname = "DBI:mysql:$db_name";              # set global $dbname 
        }
        else 
        {   # hostname changed... more tricky 
            $db_name ||= "mp3";  # revert to default
            # set global $dbname
            $dbname = "DBI:mysql:database=$db_name:host=$db_host" .  
                (defined $db_port ? ":port=$db_port" : ''); 
        } 
    } 
    else 
    {   # no changes in host or db name, but wiseguys may set port 
        if (defined $db_port) 
        {   # port on what?... (I'm probably making an error here...)
            $dieflag = 1; 
            $dieerror .= "--dbport requires --dbhost to be set\n"; 
        } 
    }

    $dbuser = $db_user if (defined $db_user);       # set global $dbuser
    $dbpass = $db_pass if (defined $db_pass);       # set global $dbpass
    $mp3_player = $player if (defined $player);     # set global $mp3_player

    if (!-x $mp3_player)
    {   # we can't run the mp3 player!  Lot of good that'll do :-)
        $dieflag = 1;
        $dieerror .= "Can't run the mp3 player:  $mp3_player\n";
    }

    if (defined $buffer)
    {   # validate positive integer
        if ($buffer < 0)
        {
            $dieflag = 1;
            $dieerror .= "--buffer must specify a number >= 0\n";
        }
        else
        {
            $play_buffer = $buffer if (defined $buffer);    # set global $play_buffer
        }
    }

    $ps = $local_ps if (defined $local_ps);

    if (! -x $ps)
    {   # can't run this 'ps', so what's the point?
        $dieflag = 1;
        $dieerror .= "cannot run ps program [$ps]\n";
    }

    # seasons don't fear the reaper
    die "\n$dieerror\n$usage" if $dieflag;   

    # start logging unless disabled
    open_LOG($logfile) unless $nolog;
}

# when parent (vigilante) receives a SIGCHLD then the child (player) has exited
# 
# parent waits to avoid zombies then calls EXIT itself, which should shut down
# cleanly.
sub REAPER
{
    LOG("Saw a SIGCHLD.");
    wait;
    LOG("Player exited.  Shutting down.");
    EXIT();
}
